src/System.Private.CoreLib/shared/System/Collections/Generic/Dictionary.cs

```diff
@@ -716,7 +728,7 @@ private void Resize(int newSize, bool forceNewHashCodes)
             Debug.Assert(_entries != null, "_entries should be non-null");
             Debug.Assert(newSize >= _entries.Length);
 
-            int[] buckets = new int[newSize];
+            _buckets = new int[newSize];
             Entry[] entries = new Entry[newSize];
```

Comment:
If the second array allocation fails, that would leave the `Dictionary` instance corrupted.  While `Dictionary` is not hardened to OOM exceptions, this is the most probable OOM spot and it would be nice to preserve the existing behavior of not corrupting the `Dictionary`.  For that purpose, you might postpone assigning the new value to `_buckets` until the final `for` loop.

Reply:
The change is because `GetBucket` uses `_buckets`

I could change it to `ref int GetBucket(uint hashCode, int[] buckets)`? Which would also likely avoid one of the extra null checks when it inlines.

Reply:
@benaadams You may assign just before the final `for` loop.

Reply:
Hmm, that would not suffice since `_fastModMultiplier` would be still updated.

Reply:
Also changed Initialize in a similar way

Reply:
Also changed Initialize in a similar way (it gets called by TrimExcess)

Reply:
`Initialize` is easy enough; but for resize will need to change the sig between 64bit and 32bit to pass it as a param

Reply:
You sent me down a rabbit hole here 😄

Reply:
Why don't we call `GetFastModMultiplier()` right after assignment to `_buckets`?  It does not have to be tangled with `ExpandPrime()`.

Reply:
Also can't ressign the hashcodes until both arrays are allocated

Reply:
Ah, nvm is in a different array

Reply:
@benaadams You may keep the original `int[] buckets = new int[newSize];` allocation where it was to (questionably) improve locality and avoid hash code recalculation if there is no memory for that array.  Only the assignment to `_buckets` should be a 'loop' earlier than in the original code.

Reply:
Done

Reply:
>  and avoid hash code recalculation

Is done in a newly assigned array so it doesn't matter; if it OOMs the fact it spent a little time doing some hashcode recalculations is probably neither here nor there.

Reply:
I agree that that probably does not matter. Thank you for going the extra mile and improving the code. I like the new version better.